### ISSUE TITLE
🐛 Bug Report: Multiple entities can use same metadata.(name/title)

### ISSUE THREAD

**Issue by C-RoblesAutodesk** [backstage/backstage#26061]
### 📜 Description

So when new entities are registered to the catalog, doesn't matter if an entity uses the same name/title as other entity, the registration action is still happening

### 👍 Expected behavior

this behavior makes a catalog confusing to read if multiple entities using the same name or title are allowed depending on which one is used to present the data in the catalog table.

### 👎 Actual Behavior with Screenshots

![Captura de pantalla 2024-08-16 a la(s) 10 23 16](https://github.com/user-attachments/assets/734feb8e-ba3a-433e-b3fa-93a51b25ffb8)

two entities using the same metadata.title

### 👟 Reproduction steps

- Register an entity using a name or title
- Register another entity using the same name or title as the previous entity.
- multiple entities having same name

### 📃 Provide the context for the Bug.

Since we use name/title in our catalog table, where these values are required to understand which content the user will expect to see, allowing same name in multiple entities

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by camilaibs** [comment]
Hello @C-RoblesAutodesk 👋🏻,
This behavior is by design, we allow it because the processor validation would be expensive.

**Comment by C-RoblesAutodesk** [comment]
it is to avoid issues with large catalogs?, for example, if we add this as part of backstage functionality, is going to crash apps where long sets of entities exist? that's why or is there another explanation? 🤔

**Comment by C-RoblesAutodesk** [comment]
Also, one problem is allowing the same entities to use the same metadata. name, if two entities share the same name, the URL is formatted like this `/catalog/default/${kind}/${entity.name}` which two entities using the same name and kind, will have the same URL

**Comment by cb-jarunmadhesh** [comment]
@C-RoblesAutodesk `metadata.name` is unique. I don't think you can register 2 entities with same name. It'll give a `detected conflicting entityRef` error.

**Comment by C-RoblesAutodesk** [comment]
That behavior @cb-jarunmadhesh is not happening for me, here is a recording where I have an entity with a unique name, then I decided to import another entity with the same name, registration happens and when looking at the entity, it just shows 1 instead of 2.
https://github.com/user-attachments/assets/00e3e28c-d88d-42f5-947a-462ce1f5a174

**Comment by cb-jarunmadhesh** [comment]
@C-RoblesAutodesk So you're saying that when you register the second entity with the same name, the first entity is being updated with the details of second. is it? Isn't that expected? 

I just noticed that `detected conflicting entityRef` error comes when I try to register 2 entities with same name with different entity providers.

**Comment by C-RoblesAutodesk** [comment]
the example here is - `user 1 `registers an entity with the `name 'a'`, `user 2` registers an entity with the `name 'a'` but `doesn't know that exists an entity with that name`, and the `first entity is updated` with details from the second, even if the details are different, looking to have control of this at the moment of registration where if the entity already exists, warning the user about this can happen, currently as you can see in the attachment, registration happens as if a new entity is registered, but at the end is not, and the data is being ingested by the first entity which uses this name.

**Comment by freben** [comment]
To clarify:

The catalog backend absolutely does not allow two output entities to have the same ref (`kind:namespace/name` triplet). That's enforced by a database uniqueness constraint and is a hard limit.

There are various ways for entities to TRY to take a given ref slot in the catalog backend. No matter the method, only the first one will win, and all others will be completely rejected. There is absolutely not any "merging" of data where a competing entity overwrites anything in a pre-existing one, even partially.

URLs are based on refs, and therefore once an entity has taken a ref, that's exactly what the URL will forever refer to, no matter what else is going on around it.

Everything you write above aligns with this, as far as I can tell at least.

As a completely separate concern from all of this, there's an import wizard that tries to analyze a location that you point to, to try to be helpful with finding problems with it before even trying to submit it to the catalog backend. That wizard may have some issue where it doesn't properly detect that IF you press submit it will lead to some collision or so. If that's the case, we could look into what that wizard might be up to.

**Comment by freben** [comment]
I'll also add that checking things (at a cost) during registration isn't a problem. That happens only once every now and then. But performing complex large-scope sweeping validation in the "hot" processing loop is a recipe for performance problems. This means that while you can be thorough during registration, that only takes you so far - the entities may change one second later because someone makes a pull request that modifies the yaml to now have a colliding title with something else. These types of "soft" errors or suboptimal states, is something that we look for separately using offline processes, without blocking users' flows.

**Comment by C-RoblesAutodesk** [comment]
Got it, so basically the import wizards do the frontend check, and even if we have the import action available when clicking on that, the backend does all the reference validation and that's why we don't see any error till we find out that our catalog is not showing the new entity right? because it enforces uniqueness? Hope I understand correctly, but why don't do this at the analysis level, or analysis not pretend to do these validations such as looking if the reference already exists?

**Comment by freben** [comment]
Yeah agreed. The _backend part_ of the analysis only looks at the entity body and sees that it's OK and reports that back to the frontend - because the backend doesn't know _what your intent is_ with that entity body. Arguably, the backend part is doing the right thing. If you were editing the original YAML file and passed that for validation, this would be perfectly correct behaviour.

However, the frontend part _does_ know that the cause of the analysis was to register a URL that will potentially lead to a collision down the line. So that additional check should be added to the `catalog-import` process. It might want to at least put up a warning panel saying "Hey this looks OK but it seems that if you go ahead this might lead to a collision because X (with clickable link) was already registered, do you really want to do this?". That would be helpful. There are some pretty complex corner cases with this (like if you have an entire tree of things that refer to each other and one of them effectively is a benign "re-registration" of an already existing thing) but it'd still be helpful to most situations I bet.

Any help appreciated here 🙏

**Comment by C-RoblesAutodesk** [comment]
Great @freben,

We appreciate the opportunity to contribute a solution to Backstage for this issue. Our team will proceed with the following steps:

1. Document several potential solutions
2. Review these solutions internally, considering:
   - Effectiveness in solving the current issue
   - Performance implications
   - Implementation complexity
   - Scalability and maintainability

Our Acceptance Criteria (AC) for this feature request are as follows:

1. Comprehensive analysis of the current system and its limitations
2. Proposed solution that improves conflict detection without significant performance impact
3. Prototype or detailed design for implementing the proposed solution
4. Documentation of potential challenges and strategies to address them
5. Recommendation for next steps, including any required changes to the existing architecture

Given that this is a feature request, we understand that the Backstage team will need to validate these changes. To facilitate this process, we'd like to know:

1. Are there any specific guidelines or criteria for feature contributions that we should be aware of?
2. What is the preferred format for presenting our findings and proposed solution?
3. Is there a particular review process we should prepare for?
4. Are there any additional considerations or requirements you'd like us to keep in mind during our analysis and design phase?

Your insights on these points would be invaluable in ensuring our contribution aligns well with Backstage's development practices and standards. We're committed to delivering a high-quality solution that integrates seamlessly with the existing architecture.

Please let us know if you have any additional thoughts or requirements that could help us in finding the optimal solution. We're looking forward to collaborating on this improvement for Backstage.

**Comment by camilaibs** [comment]
Hello @C-RoblesAutodesk, is your team still willing to collaborate with us? It would be greatly appreciated if you could 🙂
